File: xo-digest/webhook/sanity/worker.py
// Cloudflare Worker: /webhook/sanity for xo-digest.com
export default {
  async fetch(request, env, ctx) {
    if (request.method !== 'POST') {
      return new Response('Method Not Allowed', { status: 405 });
    }

    const webhookSecret = env.SANITY_WEBHOOK_SECRET || 'xo_sanity_default';
    const headers = Object.fromEntries(request.headers.entries());
    const rawBody = await request.text();


    const cryptoKey = await crypto.subtle.importKey(
      'raw',
      new TextEncoder().encode(webhookSecret),
      { name: 'HMAC', hash: 'SHA-256' },
      false,
      ['verify']
    );

    const signatureHeader = headers['x-sanity-signature'];
    if (!signatureHeader) {
      return new Response('Missing signature header', { status: 400 });
    }

    const valid = await crypto.subtle.verify(
      'HMAC',
      cryptoKey,
      Uint8Array.from(atob(signatureHeader), c => c.charCodeAt(0)),
      new TextEncoder().encode(rawBody)
    );

    if (!valid) {
      return new Response('Invalid signature', { status: 401 });
    }

    try {
      const payload = JSON.parse(rawBody);
      const type = payload?._type || 'unknown';
      const id = payload?._id || 'none';

      console.log(`[Webhook] Received ${type} update: ${id}`);

      // Relay to main handler or fallback with logging
      const relayRes = await fetch(env.XO_DIGEST_RELAY_URL, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: rawBody
      });

      if (!relayRes.ok && env.XO_DIGEST_RELAY_FALLBACK_URL) {
        // Fallback relay
        await fetch(env.XO_DIGEST_RELAY_FALLBACK_URL, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: rawBody
        });

        // Optional: Discord log hook
        if (env.XO_DIGEST_LOG_HOOK) {
          await fetch(env.XO_DIGEST_LOG_HOOK, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
              content: `⚠️ Relay failure on ${new Date().toISOString()}`,
              embeds: [{
                title: "Relay Status",
                description: `Relay to \`${env.XO_DIGEST_RELAY_URL}\` failed, fallback triggered.`,
                color: 0xffcc00,
                footer: { text: "XO Digest Webhook" },
                timestamp: new Date().toISOString()
              }]
            })
          });
        }
        // Optional: Telegram alert
        if (env.XO_DIGEST_TELEGRAM_HOOK_URL) {
          await fetch(env.XO_DIGEST_TELEGRAM_HOOK_URL, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
              text: `⚠️ *XO Digest Relay Fallback Triggered*\n\nRelay to \`${env.XO_DIGEST_RELAY_URL}\` failed.\n\n*Payload Type:* ${type}\n*ID:* ${id}\n*Time:* ${new Date().toISOString()}`,
              parse_mode: "Markdown"
            })
          });
        }
      }

      // Optional: markdown preview (stub)
      // const markdownPreview = renderMarkdownPreview(payload); // Implement if needed

      // Optional: queue IPFS upload (mock, non-blocking)
      if (env.XO_DIGEST_IPFS_UPLOAD_URL) {
        fetch(env.XO_DIGEST_IPFS_UPLOAD_URL, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: rawBody
        }).catch(err => console.warn('[IPFS Upload Error]', err));
      }

      return new Response('✅ Received webhook', { status: 200 });
    } catch (err) {
      console.error('[Webhook Error]', err);
      return new Response('Invalid JSON', { status: 400 });
    }
  },
};

File: xo-digest/webhook/ipfs-upload/worker.js
export async function uploadToIPFS({ filename, content, markdown = false, token }) {
  const mimeType = markdown ? "text/markdown" : "text/plain";
  const file = new File([content], filename, { type: mimeType });

  const client = new NFTStorage({ token });
  const cid = await client.storeBlob(file);
  return {
    cid,
    url: `https://ipfs.io/ipfs/${cid}`,
    filename
  };
}

import { NFTStorage, File } from 'nft.storage'

const NFT_STORAGE_TOKEN = "YOUR_NFT_STORAGE_API_KEY"; // Consider using env.NFT_STORAGE_TOKEN

export default {
  async fetch(request, env, ctx) {
    if (request.method !== "POST") {
      return new Response("Only POST requests allowed", { status: 405 });
    }

    try {
      const contentType = request.headers.get("Content-Type") || "";
      if (!contentType.includes("application/json")) {
        return new Response("Expected JSON payload", { status: 400 });
      }

      const payload = await request.json();
      const { filename, content, markdown } = payload;

      if (!filename || !content) {
        return new Response("Missing required fields: filename or content", { status: 400 });
      }

      const { cid, url } = await uploadToIPFS({
        filename,
        content,
        markdown,
        token: env.NFT_STORAGE_TOKEN || NFT_STORAGE_TOKEN
      });

      return Response.json({
        status: "success",
        message: "File uploaded to IPFS",
        url,
        cid,
        filename,
      });

    } catch (err) {
      return new Response(`Error: ${err.message}`, { status: 500 });
    }
  },
};